File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/common/images/glide/util/AppModelLoader.java
package com.github.anrimian.musicplayer.ui.common.images.glide.util;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.bumptech.glide.Priority;
import com.bumptech.glide.Registry;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.Options;
import com.bumptech.glide.load.data.DataFetcher;
import com.bumptech.glide.load.model.ModelLoader;
import com.bumptech.glide.load.model.ModelLoaderFactory;
import com.bumptech.glide.load.model.MultiModelLoaderFactory;
import com.bumptech.glide.signature.ObjectKey;

import javax.annotation.Nonnull;

public abstract class AppModelLoader<Model, Data> {

    public static <Model, Data> void addModelLoader(@Nonnull Registry registry,
                                                    @NonNull Class<Model> modelClass,
                                                    @NonNull Class<Data> dataClass,
                                                    @NonNull AppModelLoader<Model, Data> modelLoader) {
        registry.prepend(modelClass, dataClass, new SimpleLoaderFactory<>(dataClass, modelLoader));
    }

    protected void cleanup() {

    }

    protected void cancel() {

    }

    @NonNull
    protected DataSource getDataSource() {
        return DataSource.LOCAL;
    }

    protected abstract Object getModelKey(Model model);

    protected abstract void loadData(Model model,
                                     @NonNull Priority priority,
                                     @NonNull DataFetcher.DataCallback<? super Data> callback);

    private static class SimpleLoaderFactory<Model, Data> implements ModelLoaderFactory<Model, Data> {

        private final Class<Data> dataClass;
        private final AppModelLoader<Model, Data> appModelLoader;

        public SimpleLoaderFactory(Class<Data> dataClass, AppModelLoader<Model, Data> appModelLoader) {
            this.dataClass = dataClass;
            this.appModelLoader = appModelLoader;
        }

        @NonNull
        @Override
        public ModelLoader<Model, Data> build(@NonNull MultiModelLoaderFactory multiFactory) {
            return new SimpleModelLoader<>(dataClass, appModelLoader);
        }

        @Override
        public void teardown() {

        }

    }

    private static class SimpleModelLoader<Model, Data> implements ModelLoader<Model, Data> {

        private final Class<Data> dataClass;
        private final AppModelLoader<Model, Data> appModelLoader;

        private SimpleModelLoader(Class<Data> dataClass, AppModelLoader<Model, Data> appModelLoader) {
            this.dataClass = dataClass;
            this.appModelLoader = appModelLoader;
        }

        @Nullable
        @Override
        public LoadData<Data> buildLoadData(@NonNull Model model,
                                            int width,
                                            int height,
                                            @NonNull Options options) {
            return new LoadData<>(
                    new ObjectKey(appModelLoader.getModelKey(model)),
                    new SimpleDataFetcher<>(model, dataClass, appModelLoader)
            );
        }

        @Override
        public boolean handles(@NonNull Model model) {
            return true;
        }
    }

    private static class SimpleDataFetcher<Model, Data> implements DataFetcher<Data> {

        private final Model model;
        private final Class<Data> dataClass;
        private final AppModelLoader<Model, Data> appModelLoader;

        public SimpleDataFetcher(Model model,
                                 Class<Data> dataClass,
                                 AppModelLoader<Model, Data> appModelLoader) {
            this.model = model;
            this.dataClass = dataClass;
            this.appModelLoader = appModelLoader;
        }

        @Override
        public void loadData(@NonNull Priority priority, @NonNull DataCallback<? super Data> callback) {
            appModelLoader.loadData(model, priority, callback);
        }

        @Override
        public void cleanup() {

        }

        @Override
        public void cancel() {

        }

        @NonNull
        @Override
        public Class<Data> getDataClass() {
            return dataClass;
        }

        @NonNull
        @Override
        public DataSource getDataSource() {
            return DataSource.LOCAL;
        }

    }

}

File: app/src/main/java/com/github/anrimian/musicplayer/ui/common/images/glide/util/AppModelLoader.kt
package com.github.anrimian.musicplayer.ui.common.images.glide.util

import com.bumptech.glide.Priority
import com.bumptech.glide.Registry
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.Options
import com.bumptech.glide.load.data.DataFetcher
import com.bumptech.glide.load.model.ModelLoader
import com.bumptech.glide.load.model.ModelLoaderFactory
import com.bumptech.glide.load.model.MultiModelLoaderFactory
import com.bumptech.glide.signature.ObjectKey

abstract class AppModelLoader<Model : Any, Data> {

    companion object {
        fun <Model : Any, Data> addModelLoader(
            registry: Registry,
            modelClass: Class<Model>,
            dataClass: Class<Data>,
            modelLoader: AppModelLoader<Model, Data>
        ) {
            registry.prepend(
                modelClass,
                dataClass,
                SimpleLoaderFactory(dataClass, modelLoader)
            )
        }
    }

    protected open fun cleanup() {}

    protected open fun cancel() {}

    protected open fun getDataSource(): DataSource {
        return DataSource.LOCAL
    }

    protected abstract fun getModelKey(model: Model): Any

    protected abstract fun loadData(
        model: Model,
        priority: Priority,
        callback: DataFetcher.DataCallback<in Data>
    )

    private class SimpleLoaderFactory<Model : Any, Data>(
        private val dataClass: Class<Data>,
        private val appModelLoader: AppModelLoader<Model, Data>
    ) : ModelLoaderFactory<Model, Data> {

        override fun build(multiFactory: MultiModelLoaderFactory): ModelLoader<Model, Data> {
            return SimpleModelLoader(dataClass, appModelLoader)
        }

        override fun teardown() {}
    }

    private class SimpleModelLoader<Model : Any, Data>(
        private val dataClass: Class<Data>,
        private val appModelLoader: AppModelLoader<Model, Data>
    ) : ModelLoader<Model, Data> {

        override fun buildLoadData(
            model: Model,
            width: Int,
            height: Int,
            options: Options
        ): ModelLoader.LoadData<Data> {
            return ModelLoader.LoadData(
                ObjectKey(appModelLoader.getModelKey(model)),
                SimpleDataFetcher(model, dataClass, appModelLoader)
            )
        }

        override fun handles(model: Model): Boolean {
            return true
        }
    }

    private class SimpleDataFetcher<Model : Any, Data>(
        private val model: Model,
        private val dataClass: Class<Data>,
        private val appModelLoader: AppModelLoader<Model, Data>
    ) : DataFetcher<Data> {

        override fun loadData(
            priority: Priority,
            callback: DataFetcher.DataCallback<in Data>
        ) {
            appModelLoader.loadData(model, priority, callback)
        }

        override fun cleanup() {
            appModelLoader.cleanup()
        }

        override fun cancel() {
            appModelLoader.cancel()
        }

        override fun getDataClass(): Class<Data> {
            return dataClass
        }

        override fun getDataSource(): DataSource {
            return appModelLoader.getDataSource()
        }
    }
}